lib/ansible/plugins/action/win_copy.py

```diff
@@ -1,29 +1,522 @@
+        # characters in the filename.
+        for directory in directories:
+            directory_path = to_bytes(directory['src'], errors='surrogate_or_strict')
+            archive_path = to_bytes(directory['dest'], errors='surrogate_or_strict')
+            zip_file.write(directory_path, archive_path, zipfile.ZIP_DEFLATED)
+
+        for file in files:
+            file_path = to_bytes(file['src'], errors='surrogate_or_strict')
+            archive_path = to_bytes(file['dest'], errors='surrogate_or_strict')
+            zip_file.write(file_path, archive_path, zipfile.ZIP_DEFLATED)
+
+        return zip_file_path
+
+    def _remove_tempfile_if_content_defined(self, content, content_tempfile):
+        if content is not None:
+            os.remove(content_tempfile)
+
+    def _create_directory(self, dest, source_rel, task_vars):
+        dest_path = self._connection._shell.join_path(dest, source_rel)
+        file_args = self._task.args.copy()
+        file_args.update(
+            dict(
+                path=dest_path,
+                state="directory"
+            )
+        )
+        file_args.pop('content', None)
+
+        file_result = self._execute_module(module_name='file', module_args=file_args, task_vars=task_vars)
+        return file_result
+
+    def _copy_single_file(self, local_file, dest, source_rel, task_vars):
+        if self._play_context.check_mode:
+            module_return = dict(changed=True)
+            return module_return
+
+        # copy the file across to the server
+        tmp_path = self._make_tmp_path()
+        tmp_src = self._connection._shell.join_path(tmp_path, 'source')
+        self._transfer_file(local_file, tmp_src)
+
+        copy_args = self._task.args.copy()
+        copy_args.update(
+            dict(
+                dest=dest,
+                src=tmp_src,
+                original_basename=source_rel,
+                mode="single"
+            )
+        )
+        copy_args.pop('content', None)
+
+        copy_result = self._execute_module(module_name="copy", module_args=copy_args, task_vars=task_vars)
+        self._remove_tmp_path(tmp_path)
+
+        return copy_result
+
+    def _copy_zip_file(self, dest, files, directories, task_vars):
+        # create local zip file containing all the files and directories that
+        # need to be copied to the server
+        try:
+            zip_file = self._create_zip_tempfile(files, directories)
+        except Exception as e:
+            module_return = dict(
+                changed=False,
+                failed=True,
+                msg="failed to create tmp zip file: %s" % to_text(e),
+                exception=traceback.format_exc()
+            )
+            return module_return
+
+        zip_path = self._loader.get_real_file(zip_file)
+
+        if self._play_context.check_mode:
+            module_return = dict(changed=True)
+            os.remove(zip_path)
+            os.removedirs(os.path.dirname(zip_path))
```

Comment:
This should probably be just `rmdir`- otherwise you could end up inadvertently deleting someone's tempdir if it were empty...